File: src/main/java/com/flanks255/simplylight/blocks/RotatableLamp.java
package com.flanks255.simplylight.blocks;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;

@SuppressWarnings("deprecation")
public abstract class RotatableLamp extends LampBase implements SimpleWaterloggedBlock {
    public RotatableLamp(Properties props) {
        super(props);
        registerDefaultState(getStateDefinition().any().setValue(BlockStateProperties.WATERLOGGED, false));
    }
    public VoxelShape DOWN;
    public VoxelShape UP;
    public VoxelShape NORTH;
    public VoxelShape SOUTH;
    public VoxelShape WEST;
    public VoxelShape EAST;

    
    @Override
    public VoxelShape getShape(BlockState blockState,  BlockGetter world,  BlockPos blockPos,  CollisionContext context) {
        Direction facing = blockState.getValue(BlockStateProperties.FACING);
        //D-U-N-S-W-E
        return switch (facing.get3DDataValue()) {
            case 0 -> DOWN;
            default -> UP;
            case 2 -> NORTH;
            case 3 -> SOUTH;
            case 4 -> WEST;
            case 5 -> EAST;
        };
    }

    @Override
    public boolean canPlaceLiquid(Player player,  BlockGetter blockGetter,  BlockPos blockPos,  BlockState blockState,  Fluid fluid) {
        return true;
    }

    
    @Override
    public FluidState getFluidState(BlockState blockState) {
        return blockState.getValue(BlockStateProperties.WATERLOGGED)? Fluids.WATER.getSource(false) : super.getFluidState(blockState);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(BlockStateProperties.FACING, BlockStateProperties.WATERLOGGED);
    }

    
    @Override
    public BlockState getStateForPlacement(BlockPlaceContext context) {
        boolean waterlogged = context.getLevel().getFluidState(context.getClickedPos()).getType() == Fluids.WATER;
        return defaultBlockState().setValue(BlockStateProperties.FACING, context.getClickedFace()).setValue(BlockStateProperties.WATERLOGGED, waterlogged);
    }

    @Override
    
    public BlockState updateShape(BlockState blockState,  Direction direction,  BlockState blockState2,  LevelAccessor levelAccessor,  BlockPos blockPos,  BlockPos blockPos2) {
        if (blockState.getValue(BlockStateProperties.WATERLOGGED))
            levelAccessor.scheduleTick(blockPos, Fluids.WATER, Fluids.WATER.getTickDelay(levelAccessor));
        return super.updateShape(blockState, direction, blockState2, levelAccessor, blockPos, blockPos2);
    }

    
    @Override
    public BlockState mirror(BlockState pState, Mirror pMirror) {
        return pState.rotate(pMirror.getRotation(pState.getValue(BlockStateProperties.FACING)));
    }

    
    @Override
    public BlockState rotate(BlockState pState, Rotation pRotation) {
        return pState.setValue(BlockStateProperties.FACING, pRotation.rotate(pState.getValue(BlockStateProperties.FACING)));
    }
}

File: src/main/kotlin/com/flanks255/simplylight/blocks/RotatableLamp.kt
package com.flanks255.simplylight.blocks

import net.minecraft.block.properties.PropertyDirection
import net.minecraft.block.state.BlockFaceShape
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.util.EnumFacing
import net.minecraft.util.Mirror
import net.minecraft.util.Rotation
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

@Suppress("OverridingDeprecatedMember")

abstract class RotatableLamp(name: String): LampBase(name) {


    abstract val AABB_UP : AxisAlignedBB
    abstract val AABB_DOWN : AxisAlignedBB
    abstract val AABB_EAST : AxisAlignedBB
    abstract val AABB_WEST : AxisAlignedBB
    abstract val AABB_NORTH : AxisAlignedBB
    abstract val AABB_SOUTH : AxisAlignedBB

    companion object {
        val FACING : PropertyDirection = PropertyDirection.create("facing")
    }

    override fun getBlockFaceShape(worldIn: IBlockAccess?, state: IBlockState?, pos: BlockPos?, face: EnumFacing?): BlockFaceShape {
        return BlockFaceShape.UNDEFINED
    }

    private fun getMyBoundingBox(state: IBlockState?): AxisAlignedBB {
        return when(state?.getValue(FACING)){
            EnumFacing.DOWN-> AABB_DOWN
            EnumFacing.UP-> AABB_UP
            EnumFacing.EAST-> AABB_EAST
            EnumFacing.WEST-> AABB_WEST
            EnumFacing.NORTH-> AABB_NORTH
            EnumFacing.SOUTH-> AABB_SOUTH
            null->AABB_UP
        }
    }

    override fun getBoundingBox(state: IBlockState?, source: IBlockAccess?, pos: BlockPos?): AxisAlignedBB {
        return getMyBoundingBox(state)
    }

    override fun getCollisionBoundingBox(blockState: IBlockState?, worldIn: IBlockAccess?, pos: BlockPos?): AxisAlignedBB? {
        return getMyBoundingBox(blockState)
    }

    @SideOnly(Side.CLIENT)
    override fun isOpaqueCube(state: IBlockState?): Boolean = false

    override fun isFullCube(state: IBlockState?): Boolean = false

    override fun getStateForPlacement(worldIn: World, pos: BlockPos, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, meta: Int, placer: EntityLivingBase): IBlockState {
        //button placement: defaultState.withProperty(FACING, facing)
        //piston placement: defaultState.withProperty(FACING, EnumFacing.getDirectionFromEntityLiving(pos, placer))
        return defaultState.withProperty(FACING, facing)
    }

    override fun canPlaceBlockOnSide(worldIn: World, pos: BlockPos, side: EnumFacing): Boolean {
        val shape = worldIn.getBlockState(pos.offset(side.opposite)).getBlockFaceShape(worldIn,pos,side)
        return shape == BlockFaceShape.SOLID || shape == BlockFaceShape.CENTER
    }

    override fun createBlockState(): BlockStateContainer {
        return BlockStateContainer(this, FACING)
    }

    override fun getMetaFromState(state: IBlockState?): Int {
        return when(state?.getValue(FACING)){
            EnumFacing.UP-> 0
            EnumFacing.DOWN-> 1
            EnumFacing.EAST-> 2
            EnumFacing.WEST-> 3
            EnumFacing.NORTH-> 4
            EnumFacing.SOUTH-> 5
            null->0
        }
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        return defaultState.withProperty(FACING, when(meta){
            0->EnumFacing.UP
            1->EnumFacing.DOWN
            2->EnumFacing.EAST
            3->EnumFacing.WEST
            4->EnumFacing.NORTH
            5->EnumFacing.SOUTH
            else->EnumFacing.UP
        })
    }

    override fun withRotation(state: IBlockState, rot: Rotation): IBlockState {
        return state.withProperty(FACING, rot.rotate(state.getValue(FACING)))
    }

    override fun withMirror(state: IBlockState, mirrorIn: Mirror): IBlockState {
        return state.withRotation(mirrorIn.toRotation(state.getValue(FACING)))
    }

    override fun setCreativeTab(tab: CreativeTabs): LampBase {
        super.setCreativeTab(tab)
        return this
    }

}